src/connector/src/sink/file_sink/batching_log_sink.rs

```diff
@@ -0,0 +1,147 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use async_trait::async_trait;
+
+use crate::sink::file_sink::opendal_sink::OpenDalSinkWriter;
+use crate::sink::log_store::{LogStoreReadItem, TruncateOffset};
+use crate::sink::{LogSinker, Result, SinkLogReader};
+
+/// `BatchingLogSinker` is used for a commit-decoupled sink that supports cross-barrier batching.
+/// Currently, it is only used for file sinks, so it contains an `OpenDalSinkWriter`.
+/// When the file sink writer completes writing a file, it truncates the corresponding `chunk_id` in the log store.
+pub struct BatchingLogSinker {
+    writer: OpenDalSinkWriter,
+}
+
+impl BatchingLogSinker {
+    /// Create a log sinker with a file sink writer.
+    pub fn new(writer: OpenDalSinkWriter) -> Self {
+        BatchingLogSinker { writer }
+    }
+}
+
+#[async_trait]
+impl LogSinker for BatchingLogSinker {
+    async fn consume_log_and_sink(self, log_reader: &mut impl SinkLogReader) -> Result<!> {
+        let mut sink_writer = self.writer;
+        #[derive(Debug)]
+        enum LogConsumerState {
+            /// Mark that the log consumer is not initialized yet
+            Uninitialized,
+
+            /// Mark that a new epoch has begun.
+            EpochBegun { curr_epoch: u64 },
+
+            /// Mark that the consumer has just received a barrier
+            BarrierReceived { prev_epoch: u64 },
+        }
+
+        let mut state = LogConsumerState::Uninitialized;
+        loop {
+            let (epoch, item): (u64, LogStoreReadItem) = log_reader.next_item().await?;
+            if let LogStoreReadItem::UpdateVnodeBitmap(_) = &item {
+                match &state {
+                    LogConsumerState::BarrierReceived { prev_epoch } => {
+                        // we need to force to finish the batch here. Otherwise, there can be data loss because actor can be dropped and rebuilt during scaling.
+                        if let Some(committed_chunk_id) = sink_writer.commit().await? {
+                            // If epoch increased, we first need to truncate the previous epoch.
+                            if epoch > *prev_epoch {
+                                log_reader
+                                    .truncate(TruncateOffset::Barrier { epoch: *prev_epoch })?;
+                            }
+
+                            log_reader.truncate(TruncateOffset::Chunk {
+                                epoch: (epoch),
+                                chunk_id: (committed_chunk_id),
+                            })?
```

Comment:
I don't think we need to truncate on epoch chunk id here because all the writes are associated with data in prev epoch.